### ISSUE TITLE
Consistent mouse and key event handling

### ISSUE THREAD

**Issue by eruvanos** [pythonarcade/arcade#1822]
From [Discord](https://discord.com/channels/458662222697070613/851382569462267955/1118225793974288434)
```
There are established patterns from other UI libraries and the web which can be copied unless we have good reason to deviate.

Mouse events go to the widget when overlapping the widget's bounds, keyboard events go to the widget when it has "focus"
```

Revisit the event logic and check if these changes are applicable.

- [ ] widgets drop mouse events and do not pass them further, if they are out of bounce
- [ ] only focused widget receives text events
- [ ] one widget can be focused per UIManager

**Comment by pushfoo** [comment]
I discussed expanding pyglet's keyboard combo handling with As_bestos on the pyglet Discord [at the start of this thread](https://discord.com/channels/438622377094414346/1149618273944608839/1149635610517188618) and [some of its later messages](https://discord.com/channels/438622377094414346/1149618273944608839/1149635791014862878).

tl;dr: Expanding the hotkey dict approach pyglet uses for Windows and Linux into a "layer" model with a push / pop wrapper would allow for customization and easier hotkey handling.

Recap:

* It may be possible to expand hotkey handling by replacing the default handling structures with `Mapping[Tuple[FrozenSet[int], int], T]` on Windows and X-based *NIX DEs
  * Mac gets more complicated
  * Using a `FrozenSet` rather than the current tuple keys has multiple benefits:
    * Modifiers are no longer limited to Ctrl and whether shift is pressed
    * Can also use arbitrary non-modifier keys as modifiers
    * Could be expanded to non-key modifiers combinations
* It would also be worthwhile to allowing stacking of "layers" that can be pushed and popped 
  * This would allow enabling shortcuts only in certain contexts
  * The correct model for disabling hotkeys in a pushed layer / context is unclear, ie for modals which block input other than button nav
  * Caching combinations / lookups may be a good idea, but update behavior is unclear
  * "I wish push_handlers() had a name for the set, so you could pop the set specifically"
  * We should look into the existing `push_handlers` functionality

**Comment by pushfoo** [comment]
#1894 will help get us ready to set up cross-platform hotkeys associated with UI elements. The thread recapped above may also have mentioned an idea of providing a UI hotkey binding helper. For example, maybe something like this?

```python
# the tuples could be converted to frozensets internally, which would also
# allow further abstraction of binding action + opposite pairs at the same time
# or other operations by combining tuples / lists.
next_tab_button.bind_hotkey([key.MOD_ACCEL], key.TAB)
prev_tab_button.bind_hotkey([key.MOD_ACCEL, key.MOD_SHIFT], key.TAB)

```
